### PR TITLE
Support decoding to Collection and Map subclasses

### PR DESCRIPTION
**Pull request by jyemin**
JAVA-4742

The main complication to adding this feature is that the existing IterableCodec is not able to extend a class with the parameterized types `<T, C extends Collection<T>>`.  The same goes for MapCodec extending a class with the parameterized types `<T, M extends Map<String, T>>`.  The type system prevents this in the face of the default behavior of using ArrayList and HashMap, respectively, for the existing public constructors.  Also, since Iterable does not have an `add` method, it's problematic to add a decode method that supports parameterization with Iterable.

I tried a bunch of experiments to get these codecs to work with this new feature, failed, and went with the following approach:

- Revert the recent changes to IterableCodec and MapCodec that added support for parameterization.  Since they are part of the API they can't be removed, but they are deprecated.
- Add new CollectionCodec and MapCodecV2 codecs that don't suffer from the same problems
- Add CollectionCodecProvider and use it in all the default registries ahead of IterableCodecProvider (which has to remain to support case where it actually _is_ an Iterable that is not a Collection.  Without this we can't encode anything created by our own `com.mongodb.internal.Iterables` class, for example. )
- Create instances of MapCodecV2 instead of MapCodec in MapCodecProvider

### CODE REVIEW

File: bson/src/main/org/bson/codecs/IterableCodecProvider.java [hunk 1]
```diff
@@ -27,7 +27,9 @@
  * A {@code CodecProvider} for classes than implement the {@code Iterable} interface.
  *
  * @since 3.3
+ * @deprecated Prefer {@link CollectionCodecProvider}
  */
+@Deprecated
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/IterableCodecProvider.java:32]: As we need this for our own driver code - should it actually be deprecated?

**Review by rozza — CHANGES_REQUESTED** [review]

File: bson/src/main/org/bson/codecs/IterableCodecProvider.java (hunk 1 above)
Comment (jyemin) [reply bson/src/main/org/bson/codecs/IterableCodecProvider.java:32]: Maybe not.  I'll think about it more.

**Comment by rozza** [comment]
I'm of the opinion that if we require the IterableCodec / IterableCodecProvider it should not be deprecated but the suggestion to use the CollectionCodecProvider instead works.  

If we really want to deprecate it, it might be better to do so it in a separate ticket and refactor our code that requires it and remove it from the default registry (and accept it might break some niche usecases).

File: bson/src/main/org/bson/codecs/MapCodecV2.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * @param <M> the actual type of the Map, e.g. {@code NavigableMap<String, Object>}
+ */
+@SuppressWarnings("rawtypes")
+final class MapCodecV2<M extends Map<String, Object>> extends AbstractMapCodec<Object, M>
+        implements OverridableUuidRepresentationCodec<M>, Parameterizable {
+
+    private static final CodecRegistry DEFAULT_REGISTRY = fromProviders(asList(new ValueCodecProvider(), new BsonValueCodecProvider(),
+            new DocumentCodecProvider(), new CollectionCodecProvider(), new MapCodecProvider()));
+    private static final BsonTypeClassMap DEFAULT_BSON_TYPE_CLASS_MAP = new BsonTypeClassMap();
```
Comment (stIncMale) [inline bson/src/main/org/bson/codecs/MapCodecV2.java:53]: Both are unused.

File: bson/src/main/org/bson/codecs/MapCodecV2.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    private final BsonTypeCodecMap bsonTypeCodecMap;
+    private final CodecRegistry registry;
+    private final Transformer valueTransformer;
+    private final UuidRepresentation uuidRepresentation;
+
+    /**
+     /**
```
Comment (stIncMale) [inline bson/src/main/org/bson/codecs/MapCodecV2.java:60]: One of these lines is not needed.

File: bson/src/test/unit/org/bson/codecs/CollectionCodecSpecification.groovy [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+import static org.bson.UuidRepresentation.C_SHARP_LEGACY
+import static org.bson.UuidRepresentation.JAVA_LEGACY
+import static org.bson.UuidRepresentation.PYTHON_LEGACY
+import static org.bson.UuidRepresentation.STANDARD
+import static org.bson.UuidRepresentation.UNSPECIFIED
+import static org.bson.codecs.configuration.CodecRegistries.fromCodecs
+import static org.bson.codecs.configuration.CodecRegistries.fromProviders
+import static org.bson.codecs.configuration.CodecRegistries.fromRegistries
+
+class CollectionCodecSpecification extends Specification {
```
Comment (stIncMale) [inline bson/src/test/unit/org/bson/codecs/CollectionCodecSpecification.groovy:42]: Currently these tests pass if the `else` branch in `AbstractCollectionCodec` is replaced with `throw Assertions.fail();`. It appears that we lack a test that decodes to a custom `Collection` that has a no-arg constructor. `CopyOnWriteArrayList` can be used as such a custom `Collection` in the `should decode to specified generic class` test.

Note that a similar thing is **not** applicable to `MapCodecV2Specification`, as it tests against `WeakHashMap`.

File: bson/src/test/unit/org/bson/codecs/MapCodecV2Specification.groovy [hunk 5]
```diff
@@ -64,11 +64,11 @@ import static org.bson.codecs.configuration.CodecRegistries.fromCodecs
 import static org.bson.codecs.configuration.CodecRegistries.fromProviders
 import static org.bson.codecs.configuration.CodecRegistries.fromRegistries
 
-class MapCodecSpecification extends Specification {
```
Comment (stIncMale) [inline bson/src/test/unit/org/bson/codecs/MapCodecV2Specification.groovy:67]: We still have `IterableCodecSpecification`, but we are getting rid of `MapCodecSpecification`. Shouldn't the latter also stay? It seems reasonable to delete `IterableCodecSpecification`/`MapCodecSpecification` together with `IterableCodec`/`MapCodec`, but not before the latter are deleted.

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: bson/src/main/org/bson/codecs/IterableCodec.java [hunk 6]
```diff
@@ -17,26 +17,27 @@
...
  * Encodes and decodes {@code Iterable} objects.
  *
  * @since 3.3
+ * @deprecated Prefer {@link CollectionCodecProvider}
  */
+@Deprecated
 @SuppressWarnings("rawtypes")
-public class IterableCodec extends AbstractIterableCodec<Object>
-        implements OverridableUuidRepresentationCodec<Iterable<Object>>, Parameterizable {
+public class IterableCodec implements Codec<Iterable>, OverridableUuidRepresentationCodec<Iterable> {
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply bson/src/main/org/bson/codecs/IterableCodec.java:40]: Nope, no changes are needed as we simply reverted to #864.

File: bson/src/main/org/bson/codecs/IterableCodecProvider.java (hunk 1 above)
Comment (jyemin) [reply bson/src/main/org/bson/codecs/IterableCodecProvider.java]: Done: removed deprecation.

File: driver-core/src/main/com/mongodb/MongoClientSettings.java [hunk 7]
```diff
@@ -112,7 +114,7 @@ public final class MongoClientSettings {
      * <li>{@link com.mongodb.DBRefCodecProvider}</li>
      * <li>{@link com.mongodb.DBObjectCodecProvider}</li>
      * <li>{@link org.bson.codecs.DocumentCodecProvider}</li>
-     * <li>{@link org.bson.codecs.IterableCodecProvider}</li>
+     * <li>{@link org.bson.codecs.CollectionCodecProvider}</li>
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/MongoClientSettings.java:117]: Done: added it back

File: bson/src/test/unit/org/bson/codecs/pojo/PojoCustomTest.java [hunk 8]
```diff
@@ -78,33 +78,33 @@
...
 import static org.bson.codecs.pojo.Conventions.DEFAULT_CONVENTIONS;
 import static org.bson.codecs.pojo.Conventions.NO_CONVENTIONS;
 import static org.bson.codecs.pojo.Conventions.SET_PRIVATE_FIELDS_CONVENTION;
 import static org.bson.codecs.pojo.Conventions.USE_GETTERS_FOR_SETTERS;
-import static org.bson.codecs.pojo.Conventions.CLASS_AND_PROPERTY_CONVENTION;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 
+@SuppressWarnings("deprecated")
```
[the remark this answers is not in the document]
Comment (jyemin) [reply bson/src/test/unit/org/bson/codecs/pojo/PojoCustomTest.java]: Done: fixed typo and removed redundant suppression on methods

File: bson/src/main/org/bson/codecs/CollectionCodecProvider.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = bsonTypeClassMap.hashCode();
+        result = 31 * result + (valueTransformer != null ? valueTransformer.hashCode() : 0);
+        return result;
```
[the remark this answers is not in the document]
Comment (jyemin) [reply bson/src/main/org/bson/codecs/CollectionCodecProvider.java]: No objections.  This is what IntelliJ auto-generated, though I'm not sure why.  When I deleted it and re-generated, it used the preferred approach

Done

File: bson/src/main/org/bson/codecs/MapCodecV2.java (hunk 2 above)
Comment (jyemin) [reply bson/src/main/org/bson/codecs/MapCodecV2.java]: Done: removed

File: bson/src/main/org/bson/codecs/MapCodecV2.java (hunk 3 above)
Comment (jyemin) [reply bson/src/main/org/bson/codecs/MapCodecV2.java]: Done: removed

File: bson/src/test/unit/org/bson/codecs/MapCodecV2Specification.groovy (hunk 5 above)
Comment (jyemin) [reply bson/src/test/unit/org/bson/codecs/MapCodecV2Specification.groovy]: Done: reverted delete

File: bson/src/test/unit/org/bson/codecs/CollectionCodecSpecification.groovy (hunk 4 above)
Comment (jyemin) [reply bson/src/test/unit/org/bson/codecs/CollectionCodecSpecification.groovy:43]: Done: added CopyOnWriteArrayList to the where clause of the test.  The only lack of coverage now is the exceptional conditions.  Not sure it's worth it to cover those though.

Comment (stIncMale) [reply bson/src/test/unit/org/bson/codecs/CollectionCodecSpecification.groovy:43]: 🤷‍♂️ I am fine with not having those

**Review by stIncMale — APPROVED** [review]

**Pull request merged**